File: packages/Demo/scripts/python/scalogram.py
#name: Scalogram Python
#description: CWT (Morlet wavelet) scalogram plot
#reference: https://en.wikipedia.org/wiki/Scaleogram
#language: python
#environment: scalogram
#tags: demo, viewers, hide-suggestions
#sample: sensors/ecg.csv
#input: dataframe data [Input data table]
#input: column signal {type:numerical; allowNulls:false} [Column with signal]
#input: double sampleRate = 256.0 [Signal sample rate, in Hz]
#input: int w0 = 4 [Tradeoff between time and frequency resolution]
#input: bool removeDc = TRUE [Flag to force DC component removal]
#output: graphics [Scalogram plot]

import numpy as np
import matplotlib.pyplot as plt
from obspy.imaging.cm import obspy_sequential
from obspy.signal.tf_misfit import cwt

npts = len(data.index)
dt = 1.0 / sampleRate
t = np.linspace(0.0, dt * npts, npts)
f_min = 1.0
f_max = sampleRate / 10.0

signal = data[[signal]].values.ravel()

scalogram = cwt(signal, dt, w0, f_min, f_max, wl='morlet')
if removeDc:
    signal = signal - np.mean(signal)

fig = plt.figure()
ax = fig.add_subplot(111)

x, y = np.meshgrid(t, np.logspace(np.log10(f_min), np.log10(f_max), scalogram.shape[0]))

ax.pcolormesh(x, y, np.abs(scalogram), cmap=obspy_sequential)
ax.set_xlabel("Time, s")
ax.set_ylabel("Frequency, Hz")
ax.set_yscale('log')
ax.set_ylim(f_min, f_max)
plt.show()

File: packages/Samples/scripts/r/scalogram.R
#name: Scalogram R
#description: CWT (Morlet wavelet) scalogram plot
#reference: https://en.wikipedia.org/wiki/Scaleogram
#language: r
#tags: demo, viewers
#sample: sensors/ecg.csv
#input: dataframe data [Input data table]
#input: column signal {type:numerical; allowNulls:false} [Column with signal]
#input: double sampleRate = 256.0 [Signal sample rate, in Hz]
#input: int octaves = 8 [Octaves in scale values]
#input: int voices = 16 [Voices in scale values]
#input: bool removeDc = TRUE [Flag to force DC component removal]
#output: graphics [Scalogram plot]

require(Rwave)
require(colorRamps)

# Select signal column
signal = data[[signal]]

# Remove DC
if (removeDc == TRUE) {
  signal = signal - mean(signal)
}

# Computes the scale values for a given number of octaves and voices
getScales <- function(numOctave, numVoice) {
  scale = vector(length=(numOctave * numVoice))
  scale.index = 1
  for (octave in 0:(numOctave - 1)) {
    for (voice in 0:(numVoice - 1)) {
      scale[scale.index] = 2^(octave + (voice / numVoice))
      scale.index = scale.index + 1
    }
  }
  return (scale)
}

# Computes the corresponding frequency values for a given vector of scales,
# center frequency, and sample rate
scaleToFreq <- function(scale, center, sampleRate) {
  return (center / (scale * sampleRate * 2.0))
}

# Perform CWT
coef = cwt(signal, noctave=octaves, nvoice=voices, w0=(2.0 * pi * 0.8125), twoD=TRUE, plot=FALSE)

# Plots
time = seq(0, (length(signal) - 1) / sampleRate, by=(1.0 / sampleRate))
scales = floor(scaleToFreq(getScales(octaves, voices), 0.8125, 1.0 / sampleRate) * 10) / 10
image(time, 1:dim(coef)[2], abs(coef)^2.0, yaxt="n", ylab="Frequency, Hz", xlab="Time, s",
      col=matlab.like(256))
axis(2, at=seq(1, (octaves * voices)), labels=scales)
